Feat: Wireframe 2.1 - View existing metadata

signekb opened the pull request:
## Description

<!-- 
This template covers all PRs for Seedcase, please note that if you are
submitting a PR for changes to:

a) General documentation you should delete the sections Testing, Code
Documentation, and the first part of the Author Checklist.

b) Code you should delete the second section of the Author Checklist.

Otherwise, delete any sections that don't apply.
-->

<!-- DO NOT LEAVE THIS SECTION BLANK -->

- This PR adds a simplified version of wireframe 2.1 (overview of existing data) where users can select a row and get redirected.
- This will enable users to choose to create new metadata, edit existing metadata, and upload metadata (there pages are not completed yet, so currently all buttons redirect to the same randomly chosen page.
- If a user clicks "create" they will go to `data-import` 
- If a user clicks "edit" or "upload" they will be redirected to `column-review` (all redirects should be changed later)
- Note: "Data Rows" and "Last Upload" contain the placeholder value "Cell". 

## Related Issues

See Issue #162 (this PR does not close this issue - #307 continues the work towards closing it)

## Testing

- [X] Yes

<!-- Please explain why the tests are not needed for this PR here -->

## Reviewer Focus
<!-- Please delete as appropriate: -->
This PR requires an in-depth review.

- Could the code be clearer/improved on?
- Do we need more tests to cover what this view does?
- Should I remove `-view` at the end of paths/files/functions (since view is default)?
 
<!-- Any particular section the reviewer should focus on, anywhere that would be a good place to start? -->

## Checklist

<!-- This is to help you determine if your work is ready to be reviewed, if an item is not relevant then you can mark it as done (because you have checked and found that it isn't needed) -->

For all PRs that are not general documentation

- [X] Tests accompany or reflect changes to the code
- [X] Tests ran and passed locally
- [X] Ran the linter and formatter
- [X] Build has passed locally
- [X] Relevant documentation has been updated (none)

signekb committed:
feat: :sparkles: init existing tables view

with placeholder redirections to file-upload

signekb committed:
feat: :sparkles: init view existing table template

signekb committed:
feat: add view-existing-tables to urls

signekb committed:
fix: update href to match name of view-existing-tables

signekb committed:
feat: add existing tables view to view init

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb committed:
fix: remove unused import: ColumnMetadata

signekb committed:
fix: remove import of non-existing file-upload view

legacy. this view is now called metadata_create

signekb committed:
fix: remove style ".expandable"

Legacy from other implementation

signekb committed:
fix: :sparkles: change table to div to enable the whole row being clickable/selectable

before, using table, I could only get the href to work if it was in a table cell (which was name before this change)

signekb committed:
feat: add paragraph when there is not data to view

signekb committed:
fix: Change empty message to not include "sorry"

signekb committed:
refactor: convert to new naming scheme

signekb committed:
fix: remove unused import

signekb committed:
fix: remove typo

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb committed:
feat: add that clicking row again will deselect the row

signekb committed:
fix: add that selected_metadata_id should be equal to metadata.id to change url

signekb committed:
fix: add missing </a>

signekb committed:
fix: remove old urls from merge with main

signekb committed:
fix: remove import of old view from merge with main

signekb committed:
feat: redirect to data-import when create btn is clicked

Independently of whether a row is selected or not

signekb committed:
feat: :sparkles: add error dialog when user clicks edit or upload without selecting a row

signekb committed:
style: shorten msg variable name so ruff line length passes

signekb committed:
style: align variable naming: metadata -> existing_metadata

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb commented:
@lwjohnst86 How do I update the active deployment? It's outdated? :)

signekb committed:
style: add comments for clarity

lwjohnst86 commented:
@signekb it seems to have built now. Or is it outdated still?

signekb committed:
docs: make docstring clearer

signekb committed:
test: add basic tests of rendering and all tables shown in view

signekb committed:
test: :poop: add failing tests supposed to check whether page redirects correctly

signekb commented:
@lwjohnst86 Ah, maybe I was just too impatient

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb committed:
docs: fix file docstring

signekb committed:
test: :poop: update button upload and selected row test - still fails

pchmia committed:
test: add setUp

signekb committed:
test: :white_check_mark: fix and format redirects_with_button_upload_and_selected_row

signekb committed:
fix: remove unused argument "table_id" from view

signekb committed:
refactor: restructure if statement to avoid writing error msg twice

signekb committed:
style: rename template, view, url to `project id metadata view`

signekb committed:
test: :bug: rename to `metadata_view` to fix failing tests

So they match the recently renamed view

signekb committed:
Merge branch 'main' into feat/existing-table-view

philter87 commented on sprout/templates/project-id-metadata-view.html:
```diff
@@ -0,0 +1,98 @@
+
+{% block content %}
+
+<style>
```
Nice! I assumed that styles had to be in a different file (but that was just silly of me). I will need to redo some templates

signekb replied:
@philter87 Luke prefers to keep style in a separate file, so I'll change it here instead

philter87 commented on sprout/templates/project-id-metadata-view.html:
```diff
@@ -0,0 +1,98 @@
+    padding: 8px;
+  }
+
+  .selectable:hover {
```
Nice with the hover effect!

philter87 commented on sprout/templates/project-id-metadata-view.html:
```diff
@@ -0,0 +1,98 @@
+  <!-- three buttons will submit this form: create, edit, or upload -->
+  <nav class="right-align">
+    <button type="submit" name="button_create"><i>add</i>Create Metadata</button>
+    <button type="submit" name="button_edit"><i>edit</i>Edit Metadata</button>
```
Potentially you could add the disabled attribute to the buttons based on the query parameter
```
<nav class="right-align">
  <button {% if not request.GET.selected_metadata_id %}disabled{% endif %} type="submit" name="button_create"><i>add</i>Create Metadata</button>
  <button {% if not request.GET.selected_metadata_id %}disabled{% endif %} type="submit" name="button_edit"><i>edit</i>Edit Metadata</button>
  <button type="submit" name="button_upload"><i>upload</i>Upload Data</button>
</nav>
```

![image](https://github.com/seedcase-project/seedcase-sprout/assets/5868902/1ce73b1c-245c-40e0-ac0b-6c1ff63c2de6)

But the dialog gives a bit more information. 

What do you think @lwjohnst86

signekb replied:
I actually like your version better, @philter87. I'll change it 👍

philter87 commented on sprout/templates/view-existing-tables.html:
Should this file be deleted? Maybe a result of a rename or is it a different version you want to preserve?

philter87 reviewed (commented):
Great work! 
Only one 'big' thing. I was confused about 'view-existing-tables.html'

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb committed:
fix: update data-import view in test to be minus instead of underscore

signekb committed:
fix: remove old template view_existing_tables

signekb committed:
Merge branch 'main' into feat/existing-table-view

philter87 reviewed (dismissed):
Looks good now :)

signekb committed:
feat: :alembic: change dialog to disabled for button edit and upload

as a consequence, this removes the need for an error message, so this is removed too

signekb committed:
fix: remove error msg from tests since it has been removed from view

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb committed:
fix: update tests to include new column names

signekb committed:
docs: remove removed variable (error msg) from test docstring

signekb commented:
I'm done editing now, I promise 😁

lwjohnst86 committed:
Merge branch 'main' of https://github.com/seedcase-project/seedcase-sprout into feat/existing-table-view

lwjohnst86 committed:
chore: :lipstick: keep css style within it's own file

lwjohnst86 committed:
Merge branch 'feat/existing-table-view' of https://github.com/seedcase-project/seedcase-sprout into feat/existing-table-view

lwjohnst86 commented on sprout/static/css/style.css:
```diff
@@ -28,4 +28,35 @@ dd {
   white-space: nowrap;
   display: inline-block;
   width: 85%;
+}
```
Keep CSS styling within it's own sheet.

lwjohnst86 commented on sprout/templates/project-id-metadata-view.html:
```diff
@@ -0,0 +1,53 @@
+<form id="selectedElementsForm" method="post">
+  {% csrf_token %}
+  <div class="table">
+    <div class="table-header">
```
Hmm, what was the reason to use `div` rather than `table` HTML tag? This feels more complicated. Plus my formatter/linter is complaining of an unexpected closing div, but I can't debug it.

signekb replied:
I ran into some issues when I added the expand feature showing the column metadata. But I'll just change it again and see if I can figure out the way to expand using the `table` tag.

lwjohnst86 commented on sprout/templates/project-id-metadata-view.html:
```diff
@@ -0,0 +1,53 @@
+      {% for metadata in existing_metadata %}
+      <!-- if statement that changes url depending on selection and deselection of rows -->
+      {% if request.GET.selected_metadata_id|add:'0' == metadata.id %}
+      <a style="display:block" href="{% url 'project-id-metadata-view' %}">
```
Neat way of doing this. Though the big disadvantage here is that whenever you select a new row, the page has to be rendered to a whole other page because it is a redirect. Let's keep everything within this page and not move to other pages.

signekb replied:
I changed i now, but use quite a lot of js. I can probably be optimised.

lwjohnst86 commented on sprout/templates/project-id-metadata-view.html:
```diff
@@ -0,0 +1,53 @@
+          {% endif %}
+          <div>
+            <div
+              class="table-row selectable{% if request.GET.selected_metadata_id|add:'0' == metadata.id %} selected{% endif %}"
```
Using the `table` HTML you can use CSS to highlight on hover, rather than using `div` classes for this. Unless you have really strong reasons for using `div` rather than `table`, could you switch to `table`? It will be easier to style if it is a `table` element.

signekb replied:
I have changed it back to `table`

lwjohnst86 commented on sprout/templates/project-id-metadata-view.html:
```diff
@@ -0,0 +1,53 @@
+
+  <!-- three buttons will submit this form: create, edit, or upload -->
+  <nav class="right-align">
+    <button type="submit" name="button_create"><i>add</i>Create Metadata</button>
+    <button {% if not request.GET.selected_metadata_id %}disabled{% endif %} type="submit"
+      name="button_edit"><i>edit</i>Edit Metadata</button>
+    <button {% if not request.GET.selected_metadata_id %}disabled{% endif %} type="submit"
+      name="button_upload"><i>upload</i>Upload Data</button>
```
I'd prefer that redirect happens here, not on the Django side (see how I did it in #293). So it would be an `a` element with `href` to other pages. The less we have to send to Django for small things like this, the less traffic between server and client there has to be. 

Of course, these would only be clickable after selecting a page so will need some JavaScript.

signekb replied:
As described in another comment above, I have changed it now and use js, but I think that the js code could/should be optimised

signekb committed:
refactor: remove unused form name

signekb committed:
refactor:

signekb committed:
refactor: handle redirect in template instead of view using js

signekb committed:
refactor: move table selectable styles to static

signekb committed:
refactor: change buttons to anchors to slightly simplify code

signekb committed:
docs: change button texts to sentence case

pchmia committed:
refactor: simplified js and removed tests not valid with a js solution

signekb commented:
@philter87 and I have now reviewed the code together and simplified the scripts. 
Note two changes in particular:
1. It is now not possible to deselect a row when it has been selected (we discussed this, and it might not be necessary to deselect a row after all)
2. We have removed four tests that check that the redirects when the different buttons have been clicked. This is not possible anymore with the js solution.

signekb committed:
Merge branch 'main' into feat/existing-table-view

signekb committed:
style: align to use querySelector at all relevent lines

getElementById -> querySelector

signekb committed:
style: format

signekb committed:
fix: :lipstick: thead background colour and remove bottom border

remove `!important` since it's not necessary

lwjohnst86 reviewed (approved):
Super nice and clean! Great job!
